networkx/algorithms/tests/test_euler.py

```diff
@@ -150,6 +150,22 @@ def test_has_eulerian_path_isolated_node(self):
         G.add_node(3)
         assert nx.has_eulerian_path(G)
 
+    def test_has_eulerian_path_not_weakly_connected(self):
+        G = nx.DiGraph()
+        H = nx.Graph()
+        G.add_edges_from([(0, 1), (2, 3), (3, 2)])
+        H.add_edges_from([(0, 1), (2, 3), (3, 2)])
+        assert not nx.has_eulerian_path(G)
+        assert not nx.has_eulerian_path(H)
```

Comment:
```suggestion
    @pytest.mark.parametrize('graph', (nx.Graph, nx.DiGraph))
    def test_has_eulerian_path_not_weakly_connected(self, graph):
        G = graph()
        G.add_edges_from([(0, 1), (2, 3), (3, 2)])
        assert not nx.has_eulerian_path(G)
```

I personally prefer parametrizing tests like this when possible - it makes the test itself easier to read and is usually (IMO anyway) easier to parse when there's a failure.

Ditto for the other tests.

Reply:
Oh, I didn't we could do this. Thank you! This should make the tests with multiple types of graphs more concise and easier to read. I'll make the changes and push them through.